### PR TITLE
Changes the leg IK functions to templates

### PR DESCRIPTION
**Pull request by ysims**
^

### CODE REVIEW

File: shared/utility/motion/InverseKinematics.hpp [hunk 1]
```diff
@@ -52,21 +52,167 @@ namespace utility::motion::kinematics {
...
+        }
+        const Scalar sqrLength   = length * length;
+        const Scalar sqrUpperLeg = UPPER_LEG_LENGTH * UPPER_LEG_LENGTH;
+        const Scalar sqrLowerLeg = LOWER_LEG_LENGTH * LOWER_LEG_LENGTH;
+
+        const Scalar cosKnee = (sqrUpperLeg + sqrLowerLeg - sqrLength) / (2.0 * UPPER_LEG_LENGTH * LOWER_LEG_LENGTH);
+        knee                 = std::acos(std::fmax(std::fmin(cosKnee, 1), -1));
+
+        const Scalar cosLowerLeg = (sqrLowerLeg + sqrLength - sqrUpperLeg) / (2.0 * LOWER_LEG_LENGTH * length);
+        const Scalar lowerLeg    = std::acos(std::fmax(std::fmin(cosLowerLeg, 1), -1));
```
Comment (KipHamiltons) [inline shared/utility/motion/InverseKinematics.hpp:126]:
```suggestion
        const Scalar lowerLeg    = std::acos(std::max(std::min(cosLowerLeg, 1), -1));
```

File: shared/utility/motion/InverseKinematics.hpp [hunk 2]
```diff
@@ -52,21 +52,167 @@ namespace utility::motion::kinematics {
...
+        Eigen::Matrix<Scalar, 3, 1> hipX = ankleY.cross(unitTargetLeg);
+        Scalar hipXLength                = hipX.norm();
+        if (hipXLength > 0) {
+            hipX /= hipXLength;
+        }
+        else {
+            NUClear::log<NUClear::DEBUG>(
+                "InverseKinematics::calculateLegJoints : targetLeg and ankleY parallel. This is unhandled at "
+                "the "
+                "moment. requested pose = \n",
```
Comment (KipHamiltons) [inline shared/utility/motion/InverseKinematics.hpp:143]:
```suggestion
                "the moment. requested pose = \n",
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 3]
```diff
@@ -309,7 +309,7 @@ namespace module::motion {
         const Eigen::Matrix4f right_foot = walk_engine.getFootstep().isLeftSupport() ? Hft.matrix() : Hst.matrix();
 
         const auto joints =
-            calculateLegJoints(kinematicsModel, Eigen::Affine3f(left_foot), Eigen::Affine3f(right_foot));
+            calculateLegJoints<float>(kinematicsModel, Eigen::Affine3f(left_foot), Eigen::Affine3f(right_foot));
```
[the remark this answers is not in the document]
Comment (ysims) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:312]: Yeah I was really confused by the error it was giving me like surrrrrrely you can figure that out, but nope.

File: shared/utility/motion/InverseKinematics.hpp [hunk 4]
```diff
@@ -52,21 +52,167 @@ namespace utility::motion::kinematics {
...
+        Scalar knee       = 0;
+        Scalar anklePitch = 0;
+        Scalar ankleRoll  = 0;
+
+        // Correct for input referencing the bottom of the foot
+        Eigen::Transform<Scalar, 3, Eigen::Affine> target(target_);
+        target = target.translate(Eigen::Matrix<Scalar, 3, 1>(0.0, 0.0, model.leg.FOOT_HEIGHT));
+
+        // Tci = transformation (not necessarily homogeneous) from input coordinates to calculation coordinates
+        // TODO remove this. It was due to wrong convention use
```
[the remark this answers is not in the document]
Comment (ysims) [reply shared/utility/motion/InverseKinematics.hpp:80]: I'll make an issue. I have no idea what it even is.

**Comment by KipHamiltons** [comment]
Build broke, otherwise lgtm

File: shared/utility/motion/InverseKinematics.hpp [hunk 5]
```diff
@@ -52,21 +52,167 @@ namespace utility::motion::kinematics {
...
+
+        return positions;
+    }
+
+    template <typename Scalar>
+    [[nodiscard]] std::vector<std::pair<ServoID, Scalar>> calculateLegJoints(
         const message::motion::KinematicsModel& model,
         const Eigen::Affine3f& leftTarget,
-        const Eigen::Affine3f& rightTarget);
+        const Eigen::Affine3f& rightTarget) {
```
Comment (KipHamiltons) [inline shared/utility/motion/InverseKinematics.hpp:210]: Same for the `leftTarget`
```suggestion
        const Eigen::Transform<Scalar, 3, Eigen::Affine>& rightTarget) {
```

**Review by KipHamiltons — CHANGES_REQUESTED** [review]

**Review by KipHamiltons — APPROVED** [review]
LGTM!!!

**Pull request merged**